### DISCUSSION TITLE
Why does the window.onload event not work in next.js?

### DISCUSSION THREAD

**Discussion by jangir-ritik** [vercel/next.js#56776] [App Router]
[This article](https://runthatline.com/how-to-access-window-object-with-nextjs/) shows how to read window object and add events to it in next.js. The idea is to use useEffect and add/remove events to the window object.

useEffect(() => {
const onResize = () => changeColor()

window.addEventListener('resize', onResize)

return () => {
  window.removeEventListener('resize', onResize)
}
}, [])

Problem

Resize event works but load event doesn't work. How do I know if the web page is loaded with content? I need to conditionally render A or B based on loading but how do i know if the web page is loaded in next.js?

[I've explained in another post what and why I wish to achieve it](https://stackoverflow.com/questions/77039888/how-to-create-a-splash-screen-for-next-js-v13-app-router-which-isnt-just-cosm)

**Comment by icyJoseph** [comment]
By the time React code is hydrating/rendering, is far too late to catch that event. 

Try attaching placing the event listener outside the Component, and you'll be able to see such events.

I think for what you are trying to do, you don't necessarily need browser/window events, if your "client" code within the RootLayout, is executing, you are good to go already. You might want to set a piece of state perhaps.

**Reply by jangir-ritik** [reply]
And how do I check if all the assets(media) and other components in that particular web page are loaded? 
I understand what you mean by "far too late to catch that event"
in plain js window.onLoad would be one such mechanism, how to do it in next.js?

also, I can't use loading.js and use it in suspense because there is no data fetching.
I feel I am missing some concept, please help.

**Reply by icyJoseph** [reply]
Something like:

```jsx
"use client";
import { usePathname } from "next/navigation";
import PreloaderNew from "./components/PreloaderNew";
{other irrelevent imports ...}

if (typeof window !== 'undefined') {
  // do some checks on the window/document etc
}

export default function RootLayout({ children }) {
  const pathname = usePathname();
```

**Reply by jangir-ritik** [reply]
```
"use client";
import "./globals.css";
import localFont from "@next/font/local";
import Providers from "./components/Providers";
import PreloaderLatest from "./components/PreloaderLatest";
import { useEffect, useState } from "react";

let contentLoaded = null;

if (typeof window !== "undefined") {
  window.addEventListener("load", () => {
    contentLoaded = true;
  });
}

const trap = localFont({
  ...
});

export default function RootLayout({ children }) {
  const [loading, setLoading] = useState(!contentLoaded);
  return (
    <html lang="en">
      <body
        className={`${trap.variable} chrome-no-scrollbar`}
      >
        {loading && (
          <div className="text-white bg-red-400 flex justify-center items-center w-screen h-screen">
            Webapp loading
          </div>
        )}
        {!loading && (
          <div className="text-white bg-green-400 flex justify-center items-center w-screen h-screen">
            Webapp loaded
          </div>
        )}
      </body>
    </html>
  );
}
```

this is what I did.
My idea is to define a flag 'contentLoaded' and assign it a boolean value once the content is loaded and use this variable to initiate a state variable and then conditionally render but the issue is that I've marked it as 'use client' and I am getting hydration error.

**Reply by icyJoseph** [reply]
Yes, if you set it up that way you are asking for "trouble". On the server side, the initial `loading` value will differ from the client side initial render.

I think you might have been mislead by the "tech-influencers" who at the very early point said that `use client` means something is rendered only client side. That's *not* true! Client components ARE SSR'd!

- https://github.com/reactwg/server-components/discussions/4


Try with this:

```jsx
'use client';

import { useState, useEffect } from 'react';

export default function RootLayout({ children }) {
  const [showSplash, setShowSplash] = useState(true);

  useEffect(() => {
    if (document.readyState !== 'complete') {
      const handler = () => {
        console.log('load');
        setShowSplash(false);
      };
      window.addEventListener('load', handler);

      return () => {
        window.removeEventListener('load', handler);
      };
    } else {
      const timeout = window.setTimeout(() => {
        console.log('timeout');
        setShowSplash(false);
      }, 0);
      return () => window.clearTimeout(timeout);
    }
  }, []);

  return (
    <html lang="en">
      <body className={`chrome-no-scrollbar`}>
        {showSplash ? (
          <div className="text-white bg-red-400 flex justify-center items-center w-screen h-screen">
            Webapp loading
          </div>
        ) : (
          <div className="text-white bg-green-400 flex justify-center items-center w-screen h-screen">
            Webapp loaded
          </div>
        )}
      </body>
    </html>
  );
}
```

The timeout itself is optional, if the ready state is complete you can just set the `showSplash` to false.

- https://stackblitz.com/edit/nextjs-az2mhz?file=app%2Flayout.tsx

**Reply by jangir-ritik** [reply]
I have a couple of doubts

```javascript
useEffect(() => {
    if (document.readyState !== 'complete') {
      const handler = () => {
        console.log('load');
        alert('document loaded')
        setShowSplash(false);
      };
      window.addEventListener('load', handler);

      return () => {
        window.removeEventListener('load', handler);
      };
    } else {
      const timeout = window.setTimeout(() => {
        console.log('timeout');
        alert('timed-out')
        setShowSplash(false);
      }, 0);
      return () => window.clearTimeout(timeout);
    }
  }, []);
``` 
1. Isn't it too late to listen for load event? as you mentioned earlier
> By the time React code is hydrating/rendering, is far too late to catch that event.

- Which makes me wonder, **can a component be forced to mount before the document finishes loading in react/next.js?** 
- How can I control the initial html document in next.js 13?. (I feel, if I show a minimal splash screen there until the app injects, I'll be good to go)

2. I added alerts inside handler function and the else block, locally, only the alert inside of else block ran, effectively telling me setTimeout ran but in the stackblitz link, the alert inside handler got triggered. What is causing this change of behaviour?

**Reply by jangir-ritik** [reply]
```javascript
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    const assets = document.querySelectorAll("img, video");
    const handleAssetLoad = () => {
      setLoading(false);
    };

    const assetLoadPromises = Array.from(assets).map((asset) => {
      return new Promise((resolve) => {
        asset.addEventListener("load", resolve);
        asset.addEventListener("error", resolve); // Resolve even on error to avoid potential blocking
      });
    });

    Promise.all(assetLoadPromises).then(() => {
      setLoading(false);
    });
  }, []);
``` 

could this be a loading logic?

**Reply by icyJoseph** [reply]
I meant that to do anything reliably with those events, within React code will be brittle unless you start to account for edge cases. I'll take a look at the code issues soon enough.
